docs/guide/declarative_containers.md

```diff
@@ -0,0 +1,111 @@
+# Declarative Environment Configuration
+
+jobq allows you to define your Docker environment in a declarative YAML format. Instead of writing a Dockerfile directly.
+
+## YAML Structure
+
+The YAML has the following structure.
+```yaml
+build:
+  base_image: <base_image>
+  dependencies:
+    apt: [<apt_packages>]
+    pip: [<pip_packages>]
+  volumes:
+    - <host_path>:<container_path>
+  user:
+    name: <username>
+  config:
+    env:
+      - <env_var>: <value>
+    arg:
+      - <build_arg>: <value>
+    stopsignal: <signal>
+    shell: <shell>
+  meta:
+    labels:
+      - <label_key>: <label_value>
+  workdir: <working_directory>
+  filesystem:
+    copy:
+      - <src>: <dest>
+    add:
+      - <src>: <dest>
+```
+Let us walk through an example of each of the options.
+
+The YAML file uses a `build` key as the root, under which various aspects of the Docker image are defined.
+
+First, you have to first specify the base image of the Dockerfile, for example:
+
+```yaml
+build:
+  base_image: python:3.12-slim
+```
+
+Then, you can define system level and Python dependencies.
```

Comment:
Explain in detail the types of dependencies that can be defined:

- regular package dependencies
- requirements files (prefixed with `-r`, `--requirement`)
- Wheel files in the build context (anything ending in `.whl`)
- (editable) local installs (path to the project folder, must contain a `pyproject.toml`; optionally with `-e`)

In general, it might be helpful to link to the [pip docs](https://pip.pypa.io/en/stable/cli/pip_install/), where a lot more detail can be found on the requirement specification.


As for `apt` dependencies, we should let the user know that using those requires a Debian/Ubuntu-derived base image (e.g., the official `python` images).